src/semanticlayertools/clustering/leiden.py

```diff
@@ -0,0 +1,174 @@
+        self.interslice_param = intersliceCoupling
+        self.timerange = timerange
+
+        self.outfile = os.path.join(
+            outpath,
+            f'timeclusters_{timerange[0]}-{timerange[1]}_res_{resolution}_intersl_{intersliceCoupling}.csv'
+        )
+        if os.path.isfile(self.outfile):
+            raise OSError(f'Output file at {self.outfile} exists. Aborting.')
+
+        if useGC is True:
+            edgefiles = [x for x in os.listdir(inpath) if x.endswith('_GC.net')]
+        elif useGC is False:
+            edgefiles = [x for x in os.listdir(inpath) if x.endswith('.ncol')]
+
+        self.graphDict = {}
+
+        for idx in tqdm(range(len(edgefiles)), leave=False):
+            try:
+                year = re.findall(r'\d{4}', edgefiles[idx])[0]
+            except Exception:
+                raise
+            if timerange[0] <= int(year) <= timerange[1]:
+                if useGC is True:
+                    graph = ig.Graph.Read_Pajek(os.path.join(inpath, edgefiles[idx]))
+                elif useGC is False:
+                    graph = ig.Graph.Read_Ncol(os.path.join(inpath, edgefiles[idx]))
+                self.graphDict[year] = graph
+
+        self.optimiser = la.Optimiser()
+
+        print(
+            "Graphs between "
+            f"{min(list(self.graphDict.keys()))} and "
+            f"{max(list(self.graphDict.keys()))} "
+            f"loaded in {time.time() - starttime} seconds."
+        )
+
+    def optimize(self, clusterSizeCompare: int = 1000):
+        """Optimize clusters accross time slices.
+
+        This runs the actual clustering and can be very time and memory
+        consuming for large networks. Depending on the obtained cluster results,
+        this method has to be run iteratively with varying resolution parameter.
+        Output is written to file, with filename containing chosen parameters.
+
+        The output CSV contains information on which node in which year belongs
+        to which cluster. As a first measure of returned clustering, the method
+        prints the number of clusters found above a threshold defined by
+        `clusterSizeCompare`. This does not influence the output clustering.
+
+        :param clusterSizeCompare: Threshold for `interesting` clusters
+        :type clusterSizeCompare: int
+        :returns: Tuple of output file path and list of found clusters in tuple format (node, year, cluster)
+        :rtype: tuple
+
+        .. seealso::
+           Documentation of time-layer creation routine:
+           `Leiden documentation <https://leidenalg.readthedocs.io/en/latest/multiplex.html#temporal-community-detection>`_
+        """
+        starttime = time.time()
+
+        layers, interslice_layer, _ = la.time_slices_to_layers(
+            list(self.graphDict.values()),
+            interslice_weight=self.interslice_param,
+            vertex_id_attr='name'
+        )
+        print('\tSet layers.')
+
+        partitions = [
```

Comment:
Some comments would be very useful to the reader who (like me) doesn't know what CPMVertexPartition or optimize_partition_multiplex do. Just a short description, not the documentation of these functions.

Reply:
ok will add them